src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -32,21 +32,37 @@
 #include "ray/gcs/gcs_server/store_client_kv.h"
 #include "ray/gcs/store_client/observable_store_client.h"
 #include "ray/pubsub/publisher.h"
+#include "ray/util/util.h"
 
 namespace ray {
 namespace gcs {
 
+inline std::ostream &operator<<(std::ostream &str, GcsServer::StorageType val) {
+  switch (val) {
+  case GcsServer::StorageType::IN_MEMORY:
+    return str << "StorageType::IN_MEMORY";
+  case GcsServer::StorageType::REDIS_PERSIST:
+    return str << "StorageType::REDIS_PERSIST";
+  case GcsServer::StorageType::UNKNOWN:
+    return str << "StorageType::UNKNOWN";
+  default:
+    UNREACHABLE;
+  }
+}
+
 GcsServer::GcsServer(const ray::gcs::GcsServerConfig &config,
                      instrumented_io_context &main_service)
     : config_(config),
-      storage_type_(StorageType()),
+      storage_type_(GetStorageType()),
       main_service_(main_service),
       rpc_server_(config.grpc_server_name,
                   config.grpc_server_port,
                   config.node_ip_address == "127.0.0.1",
+                  cluster_token_promise_.get_future(),
```

Comment:
The main thread will be blocked on `future.get()`, but my understanding is that the `promise.set_value()` is executed by grpc on its private executors since it is a reply callback. The alternative to future is to use the sync API, but it is actually the same under the hood: sync API calls async API, and then waits for the future anyway. 

In the deadlocks that I encountered in this change, all of them were related to when `instrumented_io_context.run()` is called (IO event loop started) relative to when the RPC is sent, and not related to using future.

Question of code complexity is valid, though.

Reply:
> The alternative to future is to use the sync API

I feel the code will be cleaner this way: we can just pass the actual cluster id around after the sync API?